ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java

```diff
@@ -880,10 +881,20 @@ void processReplicaMetadataResponse(Set<MessageInfo> missingRemoteStoreMessages,
         // it is deleted in the remote store and not deleted yet locally.
 
         // if the blob is from deprecated container, then nothing needs to be done.
-        if (replicationConfig.replicationContainerDeletionEnabled && skipPredicate != null && skipPredicate.test(messageInfo)) {
+        if (replicationConfig.replicationContainerDeletionEnabled && skipPredicate != null && skipPredicate.test(
+            messageInfo)) {
           continue;
         }
-        applyUpdatesToBlobInLocalStore(messageInfo, remoteReplicaInfo, localKey);
+        if (localMessageInfoMap == null) {
+          // This makes sure that the potentially expensive findKeys() method isn't called for a replication loop, if
+          // all the keys being processed are missing in local store.
+          localMessageInfoMap = remoteReplicaInfo.getLocalStore()
+              .findKeys(messageInfoList.stream()
+                  .map(msgInfo -> (BlobId) remoteKeyToLocalKeyMap.get(msgInfo.getStoreKey()))
```

Comment:
we can filter out those keys we already know that are now now in local store.
```
messageInfoList.stream().filter(messageInfo -> !missingRemoteStoreMessage.contains(messageInfo)).
```

Reply:
done.